sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -652,6 +650,17 @@ public static DoFnSchemaInformation getDoFnSchemaInformation(
       return withSideInputs(Arrays.asList(sideInputs));
     }
 
+    /**
+     * Returns a new {@link ParDo} {@link PTransform} that's like this {@link PTransform} but with
+     * the specified additional side inputs. Does not modify this {@link PTransform}.
+     *
+     * <p>See the discussion of Side Inputs above for more explanation.
+     */
+    public SingleOutput<InputT, OutputT> withSideInput(String tagId, PCollectionView<?> sideInput) {
+      sideInput.setTagInternalId(tagId);
```

Comment:
We shouldn't modify the input view. Rather clone it and modify the the cloned copy.

Reply:
I tried to use SerializableUtils.clone to copy but since `PCollection`, `PValueBase`  does not implement serializable we cannot clone it with this method. Cloning it with this method makes `PCollection`, `Pipeline` and couple of other fields as `null`. I tried to implement setter but for `Pipeline` is final in `PValueBase` which affects a lot of other classes as well. 

So, cloning the object and setting just the PCollection in that cloned object works but during the execution of pipeline it fails the condition in `SideInputContainer.createReaderForViews` method with an error `Can't create a SideInputReader with unknown views`.

I think we need to set all the objects which are getting null after cloning the object. Do you have any other suggestion?